caas/kubernetes/provider/pod.go

```diff
@@ -0,0 +1,199 @@
+	if cond == nil { // Doesn't have scheduling information. Should not get here
+		return "", status.Unknown, since, nil
+	} else if r, s, m := conditionHandler(
+		cond, reasonMapper(podScheduledReasonsMap, status.Allocating)); !r {
+		return m, s, cond.LastProbeTime.Time, nil
+	}
+
+	// Have the init containers run?
+	if _, cond := GetPodCondition(&pod.Status, core.PodInitialized); cond != nil {
+		r, s, m := conditionHandler(
+			cond, reasonMapper(podInitializedReasonsMap, status.Maintenance))
+		if errM, isErr := InterrogatePodContainerStatus(pod.Status.InitContainerStatuses); !r && isErr {
+			return errM, status.Error, cond.LastProbeTime.Time, nil
+		} else if !r {
+			return m, s, cond.LastProbeTime.Time, nil
+		}
+	}
+
+	// Have the containers started/finished?
+	_, cond = GetPodCondition(&pod.Status, core.ContainersReady)
+	if cond == nil {
+		return "", status.Unknown, since, nil
+	} else if r, s, m := conditionHandler(
+		cond, reasonMapper(podContainersReadyReasonsMap, status.Maintenance)); !r {
+		if errM, isErr := InterrogatePodContainerStatus(pod.Status.ContainerStatuses); isErr {
+			return errM, status.Error, cond.LastProbeTime.Time, nil
+		}
+		return m, s, cond.LastProbeTime.Time, nil
+	}
+
+	// Made it this far are we ready?
+	_, cond = GetPodCondition(&pod.Status, core.PodReady)
+	if cond == nil {
+		return "", status.Unknown, since, nil
+	} else if r, s, m := conditionHandler(
+		cond, reasonMapper(podReadyReasonMap, status.Maintenance)); !r {
+		return m, s, cond.LastProbeTime.Time, nil
+	}
+
+	// If we have made it this far then something is very wrong in the state
+	// of the pod.
+	eventList, err := events()
+	if err != nil {
+		return "", "", time.Time{}, errors.Trace(err)
+	}
+
+	statusMessage := ""
+	if count := len(eventList); count > 0 {
+		statusMessage = eventList[count-1].Message
+	}
+	return statusMessage, status.Unknown, since, nil
+}
+
+// InterrogatePodContainerStatus
+func InterrogatePodContainerStatus(containers []core.ContainerStatus) (string, bool) {
+	for _, c := range containers {
+		if c.State.Running != nil {
+			continue
+		}
+
+		if c.State.Waiting != nil {
+			m, isError := isContainerReasonError(c.State.Waiting.Reason)
+			if isError {
+				m = fmt.Sprintf("%s: %s", m, c.State.Waiting.Message)
+			}
+			return m, isError
+		}
+
+		if c.State.Terminated != nil {
+			m, isError := isContainerReasonError(c.State.Terminated.Reason)
+			if isError {
+				m = fmt.Sprintf("%s: %s", m, c.State.Terminated.Message)
+			}
+			return m, isError
+		}
+	}
+	return "", false
+}
+
+func isContainerReasonError(reason string) (string, bool) {
+	switch reason {
+	case ReasonError:
+		return "container error", true
+	case ReasonImagePull:
+		return "OCI image pull error", true
+	case ReasonCrashLoopBackoff:
+		return "crash loop backoff", true
+	case ReasonCompleted:
+		return "", false
+	default:
+		return fmt.Sprintf("unknown container reason: %s", reason), true
```

Comment:
this will look a bit funny when combined with 
fmt.Sprintf("%s: %s", m, c.State.Terminated.Message)

Reply:
Done & Done